File: go/tests/orchestration_test.go
// orchestration_test.go
//
// Verifies the CircleAI.Orchestration port (orchestration.go): AgentTask factory,
// swarm config defaults + device sizing, LocalAgentDispatcher (handler routing,
// blocked-when-unregistered, quality gate blocker/warning classification, dispose),
// and the IncidentTrigger mappings (memory entry crash/security, anomaly signal).

package circleai_test

import (
	"context"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestOrchestration_AgentTaskAndConfig(t *testing.T) {
	task := circleai.NewAgentTask(circleai.AgentRoleEngineering, "do", circleai.AgentPriorityHigh, nil)
	if task.Inputs == nil || task.ID.String() == "" || task.Role != circleai.AgentRoleEngineering {
		t.Fatalf("task = %+v", task)
	}
	def := circleai.DefaultAgentSwarmConfig()
	if def.MaxConcurrency != 4 || def.TaskTimeout != 5*time.Minute || !def.RequireReviewPassBeforeDeploy {
		t.Fatalf("default config = %+v", def)
	}
	// Wearable device -> 1 concurrent (ThermalSealed classifies as wearable).
	probe := circleai.DeviceProbe{RAMAvailableBytes: 512 << 20, CPUCores: 8, ThermalClass: circleai.ThermalSealed}
	if cfg := circleai.AgentSwarmConfigForDevice(probe); cfg.MaxConcurrency != 1 {
		t.Fatalf("wearable config concurrency = %d, want 1", cfg.MaxConcurrency)
	}
}

func TestOrchestration_DispatcherRoutingAndGate(t *testing.T) {
	d := circleai.NewLocalAgentDispatcher()
	d.RegisterHandler(circleai.AgentRoleEngineering, func(ctx context.Context, task circleai.AgentTask) (circleai.SwarmResult, error) {
		return circleai.SwarmResult{TaskID: task.ID, Role: task.Role, Status: circleai.AgentStatusPassed, Output: "done"}, nil
	})
	res, err := d.Dispatch(context.Background(), circleai.NewAgentTask(circleai.AgentRoleEngineering, "x", circleai.AgentPriorityNormal, nil))
	if err != nil || res.Status != circleai.AgentStatusPassed {
		t.Fatalf("dispatch = %+v err=%v", res, err)
	}
	// No handler for Security -> Blocked.
	blocked, _ := d.Dispatch(context.Background(), circleai.NewAgentTask(circleai.AgentRoleSecurity, "y", circleai.AgentPriorityNormal, nil))
	if blocked.Status != circleai.AgentStatusBlocked {
		t.Fatalf("unregistered role should block, got %+v", blocked)
	}

	// Quality gate: [CRITICAL]/[HIGH] block, others warn.
	gate, _ := d.RunQualityGate(context.Background(), circleai.SwarmResult{
		Issues: []string{"[CRITICAL] leak", "[high] slow", "style nit"},
	})
	if gate.Passed || len(gate.Blockers) != 2 || len(gate.Warnings) != 1 {
		t.Fatalf("gate = %+v", gate)
	}

	d.Dispose()
	if _, err := d.Dispatch(context.Background(), circleai.NewAgentTask(circleai.AgentRoleEngineering, "z", circleai.AgentPriorityNormal, nil)); err == nil {
		t.Fatalf("dispatch after dispose must error")
	}
}

func TestOrchestration_IncidentFromMemoryEntry(t *testing.T) {
	ctx := "tgn.app"
	entry := circleai.EpisodicMemoryEntry{
		ID:            circleai.NewEpisodicMemoryEntry("u", "a").ID,
		RecordedAtUTC: time.Now().UTC(),
		AppContext:    &ctx,
		Tags:          map[string]string{"crash": "true", "injection": "sqli"},
	}
	tasks := circleai.IncidentTasksFromMemoryEntry(entry)
	if len(tasks) != 2 {
		t.Fatalf("crash+security should yield 2 tasks, got %d", len(tasks))
	}
	if tasks[0].Role != circleai.AgentRoleOperations || tasks[1].Role != circleai.AgentRoleSecurity {
		t.Fatalf("task roles = %v / %v", tasks[0].Role, tasks[1].Role)
	}
	// Non-incident -> empty.
	none := circleai.IncidentTasksFromMemoryEntry(circleai.EpisodicMemoryEntry{Tags: map[string]string{"locale": "en"}})
	if len(none) != 0 {
		t.Fatalf("non-incident should yield no tasks, got %d", len(none))
	}
}

func TestOrchestration_IncidentFromAnomalySignal(t *testing.T) {
	// Below threshold -> no task.
	low := circleai.NewAnomalySignal(circleai.ThreatVectorMemoryAnomaly, 0.10, "mod", "d", nil)
	if _, ok := circleai.IncidentTaskFromAnomalySignal(low, 0.30); ok {
		t.Fatalf("below-threshold signal must not dispatch")
	}
	// High-severity vector at 0.60 -> High, bumped one rank -> Critical.
	sig := circleai.NewAnomalySignal(circleai.ThreatVectorControlFlowDrift, 0.60, "companion", "drift", map[string]string{"k": "v"})
	task, ok := circleai.IncidentTaskFromAnomalySignal(sig, 0.30)
	if !ok || task.Role != circleai.AgentRoleSecurity || task.Priority != circleai.AgentPriorityCritical {
		t.Fatalf("anomaly task = %+v ok=%v (want Security/Critical)", task, ok)
	}
	if task.Inputs["k"] != "v" || task.Inputs["vector"] != "ControlFlowDrift" {
		t.Fatalf("anomaly inputs = %+v", task.Inputs)
	}
}

File: typescript/tests/orchestration.test.ts
// orchestration.test.ts
// Verifies the CircleAI.Orchestration port: task creation, dispatcher handler
// routing + blocked default, quality-gate blocker classification, LokiOrchestrator
// swarm streaming + timeout + exception wrapping, IncidentTrigger mapping, and the
// SecurityOrchestrationBridge parallel dispatch.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  AgentRole,
  AgentPriority,
  AgentStatus,
  AgentSwarmConfig,
  LocalAgentDispatcher,
  LokiOrchestrator,
  IncidentTrigger,
  SecurityOrchestrationBridge,
  createAgentTask,
  swarmResult,
  type IAgentDispatcher,
  type SwarmResult,
  type AgentTask,
} from "../src/orchestration/index";
import {
  ThreatVector,
  DefaultSecurityWatchdog,
  type AnomalySignal,
} from "../src/security/index";
import type { EpisodicMemoryEntry } from "../src/memory/index";

describe("createAgentTask", () => {
  it("stamps a fresh id, empty inputs by default, and a createdAt", () => {
    const t = createAgentTask(AgentRole.Engineering, "fix bug", AgentPriority.Normal);
    assert.ok(t.id.length > 0);
    assert.equal(t.role, AgentRole.Engineering);
    assert.equal(t.inputs.size, 0);
    assert.ok(t.createdAt instanceof Date);
    const t2 = createAgentTask(AgentRole.Engineering, "x", AgentPriority.Low);
    assert.notEqual(t.id, t2.id);
  });
});

describe("LocalAgentDispatcher", () => {
  it("routes to a registered handler", async () => {
    const d = new LocalAgentDispatcher();
    d.registerHandler(AgentRole.Review, async (task) =>
      swarmResult(task.id, task.role, AgentStatus.Passed, "ok", [], new Date()),
    );
    const t = createAgentTask(AgentRole.Review, "review", AgentPriority.Normal);
    const r = await d.dispatchAsync(t);
    assert.equal(r.status, AgentStatus.Passed);
    assert.equal(r.output, "ok");
  });

  it("returns Blocked when no handler is registered", async () => {
    const d = new LocalAgentDispatcher();
    const t = createAgentTask(AgentRole.Security, "scan", AgentPriority.Critical);
    const r = await d.dispatchAsync(t);
    assert.equal(r.status, AgentStatus.Blocked);
    assert.match(r.output, /No handler registered for role Security/);
  });

  it("classifies [CRITICAL]/[HIGH] as blockers, else warnings", async () => {
    const d = new LocalAgentDispatcher();
    const r = swarmResult(
      "id",
      AgentRole.Review,
      AgentStatus.Passed,
      "out",
      ["[HIGH] boom", "[critical] bad", "[info] fyi", "plain"],
      new Date(),
    );
    const gate = await d.runQualityGateAsync(r);
    assert.equal(gate.passed, false);
    assert.deepEqual([...gate.blockers], ["[HIGH] boom", "[critical] bad"]);
    assert.deepEqual([...gate.warnings], ["[info] fyi", "plain"]);
  });

  it("throws after dispose", async () => {
    const d = new LocalAgentDispatcher();
    d.dispose();
    const t = createAgentTask(AgentRole.Review, "x", AgentPriority.Normal);
    await assert.rejects(() => d.dispatchAsync(t));
  });
});

describe("LokiOrchestrator", () => {
  it("streams a result per task and blocks gate failures", async () => {
    const d = new LocalAgentDispatcher();
    d.registerHandler(AgentRole.Engineering, async (task) =>
      swarmResult(
        task.id,
        task.role,
        AgentStatus.Passed,
        "done",
        task.description === "bad" ? ["[HIGH] nope"] : [],
        new Date(),
      ),
    );
    const orch = new LokiOrchestrator(d, AgentSwarmConfig.default());
    const tasks = [
      createAgentTask(AgentRole.Engineering, "good", AgentPriority.Normal),
      createAgentTask(AgentRole.Engineering, "bad", AgentPriority.Normal),
    ];
    const results: SwarmResult[] = [];
    for await (const r of orch.runSwarmAsync(tasks)) results.push(r);
    assert.equal(results.length, 2);
    assert.equal(results[0].status, AgentStatus.Passed);
    assert.equal(results[1].status, AgentStatus.Blocked);
    assert.ok(results[1].issues.includes("[HIGH] nope"));
  });

  it("wraps a dispatcher exception as a Failed result instead of breaking the stream", async () => {
    const throwing: IAgentDispatcher = {
      dispatchAsync: async () => {
        throw new Error("kaboom");
      },
      runQualityGateAsync: async (r) => ({ passed: r.issues.length === 0, blockers: [], warnings: [] }),
    };
    // Disable gate enforcement so the wrapped Failed result is yielded as-is. With
    // the default config (both require-flags true), RunSwarmAsync re-classifies any
    // gate-failing result to Blocked (LokiOrchestrator.cs lines 75-83) — so the C#
    // reference tests that assert a Failed result (e.g.
    // RunSwarmAsync_TaskExceedsTimeout_YieldsFailed) construct the orchestrator with
    // `new AgentSwarmConfig(1, …, false, false)`. We mirror that here.
    const cfg: AgentSwarmConfig = {
      ...AgentSwarmConfig.default(),
      requireReviewPassBeforeDeploy: false,
      requireSecurityPassBeforeDeploy: false,
    };
    const orch = new LokiOrchestrator(throwing, cfg);
    const results: SwarmResult[] = [];
    for await (const r of orch.runSwarmAsync([
      createAgentTask(AgentRole.Operations, "x", AgentPriority.Normal),
    ])) {
      results.push(r);
    }
    assert.equal(results.length, 1);
    assert.equal(results[0].status, AgentStatus.Failed);
    assert.match(results[0].output, /Dispatcher threw/);
  });

  it("times out a slow task", async () => {
    const slow: IAgentDispatcher = {
      dispatchAsync: (_t, signal) =>
        new Promise((resolve, reject) => {
          const timer = setTimeout(
            () => resolve(swarmResult("x", AgentRole.Review, AgentStatus.Passed, "late", [], new Date())),
            10_000,
          );
          signal?.addEventListener("abort", () => {
            clearTimeout(timer);
            reject(new Error("aborted"));
          });
        }),
      runQualityGateAsync: async () => ({ passed: true, blockers: [], warnings: [] }),
    };
    const cfg: AgentSwarmConfig = { ...AgentSwarmConfig.default(), taskTimeoutMs: 20 };
    const orch = new LokiOrchestrator(slow, cfg);
    const results: SwarmResult[] = [];
    for await (const r of orch.runSwarmAsync([
      createAgentTask(AgentRole.Review, "x", AgentPriority.Normal),
    ])) {
      results.push(r);
    }
    assert.equal(results[0].status, AgentStatus.Failed);
    assert.match(results[0].output, /timed out/);
  });
});

describe("IncidentTrigger.fromMemoryEntry", () => {
  const entry = (tags: Record<string, string>): EpisodicMemoryEntry => ({
    id: "ep1",
    recordedAtUtc: new Date("2026-01-01T00:00:00Z"),
    userText: "u",
    assistantText: "a",
    appContext: "app",
    tags,
  });

  it("returns empty for a non-incident entry", () => {
    assert.equal(IncidentTrigger.fromMemoryEntry(entry({ chat: "1" })).length, 0);
  });

  it("returns one Operations task for a crash", () => {
    const tasks = IncidentTrigger.fromMemoryEntry(entry({ crash: "1" }));
    assert.equal(tasks.length, 1);
    assert.equal(tasks[0].role, AgentRole.Operations);
    assert.equal(tasks[0].priority, AgentPriority.High);
    assert.equal(tasks[0].inputs.get("episode_id"), "ep1");
  });

  it("adds a Security task when a security tag is also present", () => {
    const tasks = IncidentTrigger.fromMemoryEntry(entry({ exception: "1", injection: "1" }));
    assert.equal(tasks.length, 2);
    assert.equal(tasks[1].role, AgentRole.Security);
    assert.equal(tasks[1].priority, AgentPriority.Critical);
  });
});

describe("IncidentTrigger.fromAnomalySignal", () => {
  const sig = (confidence: number, vector: ThreatVector): AnomalySignal => ({
    id: "s1",
    vector,
    confidence,
    affectedModule: "mod",
    description: "d",
    evidence: { k: "v" },
    detectedAt: new Date("2026-01-01T00:00:00Z"),
  });

  it("returns null below threshold", () => {
    assert.equal(IncidentTrigger.fromAnomalySignal(sig(0.2, ThreatVector.Unknown)), null);
  });

  it("maps confidence to priority and bumps high-severity vectors", () => {
    // 0.6 → High; StateCorruption is high-severity → bumped to Critical.
    const t = IncidentTrigger.fromAnomalySignal(sig(0.6, ThreatVector.StateCorruption));
    assert.ok(t);
    assert.equal(t?.role, AgentRole.Security);
    assert.equal(t?.priority, AgentPriority.Critical);
    assert.equal(t?.inputs.get("vector"), "StateCorruption");
    assert.equal(t?.inputs.get("k"), "v");
  });

  it("keeps Normal priority for a low-severity vector at 0.5", () => {
    const t = IncidentTrigger.fromAnomalySignal(sig(0.5, ThreatVector.MemoryAnomaly));
    assert.equal(t?.priority, AgentPriority.Normal);
  });
});

describe("SecurityOrchestrationBridge", () => {
  it("returns the inner watchdog response and dispatches in parallel", async () => {
    const inner = new DefaultSecurityWatchdog();
    const d = new LocalAgentDispatcher();
    let dispatched = 0;
    d.registerHandler(AgentRole.Security, async (task) => {
      dispatched++;
      return swarmResult(task.id, task.role, AgentStatus.Passed, "handled", [], new Date());
    });
    const orch = new LokiOrchestrator(d);
    const bridge = new SecurityOrchestrationBridge(inner, orch, 0.3);
    const signal: AnomalySignal = {
      id: "a1",
      vector: ThreatVector.PrivilegeEscalation,
      confidence: 0.9,
      affectedModule: "mod",
      description: "d",
      evidence: {},
      detectedAt: new Date(),
    };
    const resp = await bridge.onAnomalyDetectedAsync(signal);
    assert.ok(resp);
    // Give the fire-and-forget agent path a tick to run.
    await new Promise((r) => setTimeout(r, 5));
    assert.equal(dispatched, 1);
  });
});
